### PR TITLE
feat(debug): support creation bytecode override

### PR DESCRIPTION
**Pull request by lwedge99**

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -392,11 +401,12 @@ func (st *StateTransition) TransitionDb(refunds bool, gasBailout bool) (*Executi
...
 		// It does get incremented inside the `Create` call, after the computation
 		// of the contract's address, but before the execution of the code.
-		ret, _, st.gas, vmerr = st.evm.Create(sender, st.data, st.gas, st.value)
+		code := vm.ApplyCreationCodeOverride(st.data, st.CreationCodeOverrides)
+		ret, _, st.gas, vmerr = st.evm.Create(sender, code, st.gas, st.value)
 	} else {
 		// Increment the nonce for the next transaction
 		st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
-		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value, bailout)
+		ret, st.gas, vmerr = st.evm.CallWithCreationCodeOverrides(sender, st.to(), st.data, st.gas, st.value, bailout, st.CreationCodeOverrides)
```
Comment (ty-sentio-xyz) [inline core/state_transition.go:409]: It seems CreationCodeOverrides will not be passed on if there is another call inside this call.

For example, STATICCALL makes another `evm.StaticCall`, without `CreationCodeOverrides`:
https://github.com/sentioxyz/erigon/blob/af659e63794d0734e908af17c6c0accadb190eed/core/vm/instructions.go#L838

File: turbo/transactions/tracing.go [hunk 2]
```diff
@@ -186,7 +186,7 @@ func TraceTx(
 		stream.WriteObjectField("structLogs")
 		stream.WriteArrayStart()
 	}
-	result, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas()), refunds, false /* gasBailout */)
+	result, err := core.ApplyMessageWithCreationCodeOverrides(vmenv, message, new(core.GasPool).AddGas(message.Gas()), refunds, false /* gasBailout */, config.CreationCodeOverrides)
```
Comment (ty-sentio-xyz) [inline turbo/transactions/tracing.go:189]: How about storing this override in `vm.Config` ([link](https://github.com/sentioxyz/erigon/blob/af659e63794d0734e908af17c6c0accadb190eed/core/vm/interpreter.go#L31)), which is accessible every where, even if there are nested calls.
In this case modification can be minimum (only 3~4 files I guess).

File: eth/tracers/api.go [hunk 3]
```diff
@@ -3,18 +3,20 @@ package tracers
...
-	NoRefunds      *bool // Turns off gas refunds when tracing
-	StateOverrides *ethapi.StateOverrides
-	GasOverride    *uint64
+	Tracer                *string
+	TracerConfig          *json.RawMessage
+	Timeout               *string
+	Reexec                *uint64
+	NoRefunds             *bool // Turns off gas refunds when tracing
+	StateOverrides        *ethapi.StateOverrides
+	GasOverride           *hexutil.Big
```
Comment (ty-sentio-xyz) [inline eth/tracers/api.go:20]: Is this change to `GasOverride` directly related to the PR?

File: eth/tracers/api.go [hunk 4]
```diff
@@ -3,18 +3,20 @@ package tracers
...
+	CreationCodeOverrides map[string]hexutil.Bytes
```
Comment (ty-sentio-xyz) [inline eth/tracers/api.go:21]: I see that we are using hash of code as the key for overrides.  How about using contract address (of the one newly created), does it make more sense?

File: eth/tracers/api.go (hunk 3 above)
Comment (lwedge99) [reply eth/tracers/api.go:20]: Actually not, I will make it a separate PR.

File: core/state_transition.go (hunk 1 above)
Comment (lwedge99) [reply core/state_transition.go:409]: For other CALLs, such as `CALLCODE`, `DELEGATECALL`, it should be populated properly. But I actually miss it in `CALL`. https://github.com/sentioxyz/erigon/blob/af659e63794d0734e908af17c6c0accadb190eed/core/vm/instructions.go#L808
For `STATICCALL`, I think `CREATE` and `CREATE2` are not allowed in it, so I did not modify `opStaticCall`.

File: eth/tracers/api.go (hunk 4 above)
Comment (lwedge99) [reply eth/tracers/api.go:21]: I used to think that we are unable to get contract address before it is created, so I choose to use the hash of code as the key. But after checking a bit just now, it seems okay to directly use contract address.

File: turbo/transactions/tracing.go (hunk 2 above)
Comment (lwedge99) [reply turbo/transactions/tracing.go:189]: I thought `EVM` object is persisted between calls before and it is not feasible to store the override in it, but after looking around it seems that there will be a new `EVM` for every tracing call.

**Review by ty-sentio-xyz — APPROVED** [review]

**Pull request merged**